### PR TITLE
Major API rework, bugfixes, and updates

### PR DESCRIPTION
**Pull request by jkeifer**
* switch to django-ninja to get pydantic model validation, automatic OpenAPI docs, and modern ergonomics for defining the API
* upgrade dependencies
* switch project to pyproject.toml from setup.py
* pre-commit configuration with linting and type checking
* updates per linting and type checking
* dependency version management via pip-compile

### CODE REVIEW

**Comment by jkeifer** [comment]
One of the migration files had an invalid name (using `-`). I went ahead and fixed that, but existing deployments will need to update their migrations table correspondingly. That can be as simple as running `echo "update django_migrations set name = '0004_fix_bust_cache' where name = '0004_fix-bust-cache';" | snodas runsql -`.

**Comment by jkeifer** [comment]
Gonna go ahead and merge this. It's fairly significant rework of the app and without any (functional) tests its a bit risky. I've manually tested the API fairly heavily and it seems to be 100% functional. Some management commands being broken is possible, I haven't tested most of them.

tl;dr: testing an urgent need but too heavy of a lift to include in this go-around. I'm gonna yolo this one and if I find more changes are required I'll pull those in on other PRs.
